### PR TITLE
pulley: pack `dst`, `src1` and `src2` registers into 2 bytes

### PR DESCRIPTION
**Pull request by Kmeakin**
Pack the `dst`, `src1` and `src2` register operands of binary operators into 2 bytes. Solves the `TODO` at the top of `lib.rs`.

~This required moving the special registers into their own register class, and adding `smov` and `spush`/`spop` to interact with SP/FP.~
I held off on splitting special registers into their own `SReg` class because I'm not sure how to integrate it with ISLE/Regalloc

### CODE REVIEW

**Comment by cfallin** [comment]
Reassigned @fitzgen to review, I'm not sure how this intersects with your other plans to add instructions with large offsets etc; happy to review this myself if you'd like otherwise.

File: pulley/README.md
Comment (fitzgen) [inline pulley/README.md:51]: I was indeed thinking that the special registers would only be accessed from special instructions, but I think we can do even more here by bundling the whole function prologue and epilogue up into instructions that just take the size to decrement the stack frame by:

```
enter_frame 16
...
exit_frame 16
ret
```

where

```
enter_frame N

==>

push lr
push fp
fp = sp
sp = sp - N
```

and the opposite for `exit_frame N`.

File: pulley/src/regs.rs
Comment (fitzgen) [inline pulley/src/regs.rs:119]: Why all these changes? Seems unrelated?

File: pulley/src/lib.rs
Comment (fitzgen) [inline pulley/src/lib.rs:51]: We shouldn't need this, or any other general instructions for the special registers. They should only be manipulated by instructions that are doing other stuff as well, eg `call` will set `lr`. But these registers aren't allocated by our register allocator and are reserved for their special purpose, so we shouldn't need to have general instructions for them.

File: pulley/src/interp.rs
Comment (fitzgen) [inline pulley/src/interp.rs:556]: Could we split this out into its own PR?

**Review by fitzgen — COMMENTED** [review]
Thanks! `BinaryOperands` is exactly what I was imagining!

I'd prefer holding off on merging this until I can land the Cranelift backend for Pulley though, just so that I don't have to chase instruction set updates, and we can start evolving the backend and the instruction set together and keep them in sync.

File: pulley/src/regs.rs
Comment (Kmeakin) [reply pulley/src/regs.rs:119]: Encodes the invariant that registers are always in the range `0..32`, so indexing into the registers array can never fail. The other option would be to use the `#[rustc_layout_scalar_valid_range_end(31)]` attribute, but that is nightly-only.

Comment (fitzgen) [reply pulley/src/regs.rs:119]: Are you saying we get better codegen with `enum`s here? That's neat, I wouldn't have expected the LLVM IR that rustc emits to change.

File: pulley/README.md
Comment (Kmeakin) [reply pulley/README.md]: I'm a bit confused about the purpose of the `N` to `enter_frame`/`exit_frame`. Each `push`/`pop` will itself automatically adjust the sp by `size_of<*u8>()`. So in the case of no other registers being stored on the stack, the argument will be zero. In the case where other registers need to be saved on the stack, `xpush{32,64}` and `xpop{32,64}` will also adjust the sp by 4/8 bytes. So in either case the argument is not needed? I don't think there is any scenario where you would need to adjust the stack pointer without also reading/writing from/to the stack

Comment (fitzgen) [reply pulley/README.md]: > I'm a bit confused about the purpose of the `N` to `enter_frame`/`exit_frame`.

To allocate the stack frame for this function, including all of its regalloc spills and all that.

That said, the way that Cranelift is organized and the abstraction boundaries it has might make doing the full function prologue in a single instruction difficult. We might have to settle for two instructions:

1. a macro instruction for `push lr; push fp; fp = sp` (and the opposite for exiting the function, maybe just fold this into `ret`)
2. an instruction for allocating and deallocating stack space: `sp = sp +/- imm32`

See `gen_prologue_frame_setup`, `gen_epilogue_frame_restore`, and `gen_sp_reg_adjust` in `../pulley_shared/abi.rs` in https://github.com/bytecodealliance/wasmtime/pull/9089 (or the equivalent methods in, e.g., `../aarch64/abi.rs`). I'm thinking that each of these should produce a single macro instruction because they otherwise always generate the exact same sequence of instructions modulo immediate constant values.

The benefit of a single macro instruction, compared to a sequence of composable and more general instructions, is performance. The expensive part of an interpreter is roundtripping through the opcode loop, by using macro instructions we are amortizing that cost.

Does that make sense?

Comment (Kmeakin) [reply pulley/README.md]: Decided the extra argument was not needed after all:
https://bytecodealliance.zulipchat.com/#narrow/stream/217126-wasmtime/topic/Pulley.20registers/near/459533622

**Comment by fitzgen** [comment]
Heads up: the Pulley backend for Cranelift landed in #9089 and I think this can be rebased and landed now

**Review by fitzgen — APPROVED** [review]
Thanks!

**Pull request merged**
